RFC/src/RFC-0201_TariScript.md

```diff
@@ -183,101 +210,275 @@ pub struct TransactionOutput {
-  =& \sum(C_i + \mathrm{H}(C_i \Vert s_i).G) - \sum(C_j + \mathrm{H}(C_j \Vert s_j).G) - \sum(\mathrm{fee}.G)
+\sum(\pm k_i.G \pm L_i\sigma_i) &\stackrel?= X_s + \sum \pm L_i\sigma_i + \delta.G \\\\
+\sum \pm P_i + \sum \pm L_i\sigma_i &\stackrel?= X_s + \sum \pm L_i\sigma_i + \delta.G \\\\
+\sum \pm P_i +  &\stackrel?= X_s +  \delta.G \\\\
 \end{align}
 $$
 
-If the accounting is correct, all values will cancel
+Given the definition of the excess, \\(  \sum \pm P_i = X_s + \delta.G \\), the balance holds.
+
+If the balance does not hold for whatever reason, the transaction is rejected, as per the _status quo_. 
+
+### Kernel signature
+
+The kernel signature must also be adapted to account for the fact that the residual of the block balance is no longer 
+simply the public excess, but also includes the script hashes and script keys.
 
 $$
-  = \sum(k_i + \mathrm{H}(C_i \Vert s_i).G) - \sum(k_j + \mathrm{H}(C_j \Vert s_j).G)
+\begin{align}
+s_i &= r_i + e.\pm \bigl(k_i + k_{i'}\sigma_i \bigr) \\\\
+s_i.G &= r_i.G + e.\pm \bigl(k_i + k_{i'}\sigma_i \bigr).G \\\\
+&= R_i + e.\pm \bigl(P_i + L_i\sigma_i \bigr) \\\\
+s &= \sum s_i \\\\
+s.G  &= \sum R_i + e.\bigl( \sum \pm P_i \pm \sum L_i\sigma_i \bigr) \\\\
+&= \sum R_i + e.\bigl( X_s + \sum \pm L_i\sigma_i \bigr) \\\\
+\end{align}
 $$
 
-The sum of all the blinding factors (times G) is the definition of the standard Mimblewimble excess,
+The exercise above demonstrates that $$x_s + \sum \pm \sigma_i k_{i'}$$ signs the kernel correctly.
+
+In other words, a validating a transaction or block now checks that 
+
+$$ X_s + \sum \pm \sigma_i L_i $$ 
+
+corresponds to the kernel signature, rather than just \\( X_s \\).
+
+The same treatment extends to be block validation check. Note that since the individual kernel excesses can still be
+summed to obtain the overall block balance.
+Furthermore, at the block aggregation level, it is non-trivial to match a script hash with a UTXO, and so the
+dis-association of kernels and their outputs is maintained.
+
+### Additional consensus rules
+
+In addition to the changes given above, there are consensus rule changes for transaction and block validation.
+
+For every valid block or transaction,
+
+1. Every script hash in the kernel MUST match a script in the script section. If there are
+   duplicate non-trivial script hashes present, there MUST be a script entry for every duplicate.
+2. The script and its input data, if any, MUST execute successfully, i.e. without any errors. After execution, the stack
+   MUST have exactly one element and its value MUST be exactly zero.
+3. The Coinbase output MUST use a default script.
 
-$$ x_s\cdot G = X_s $$
+## Checking the requirements
+
+Let us evaluate this scheme against the [Additional requirements](#additional-requirements) listed earlier.
+
+### Malleability
+
+Let's say a malicious party, Bob, can modify a transaction before it enters the blockchain.
+
```

Comment:
Is it possible to perform kind of DOS attack on nodes supplying scripts which would be expensive to calculate and fail with `ContextFailure` at the end or was it accounted already? Did not find limitations on number of times scripts executed and how fee is affecting repeats. Sample script:
`Blake256 (x 1016 times) CheckHeight(100_000)`

Reply:
as a solution script might allow assertions which would lead to ContextFailure only at the start of the script.

Reply:
There are no loops, and there's a script size limit; so scripts cannot be that expensive.